File: backend/nodejs/networkUtils.js
import { deviceDiscovery } from './utils/deviceDiscovery';
import { portScanner } from './utils/portScanner';
import { networkMetrics } from './utils/networkMetrics';
import { powerManager } from './utils/powerManager';

export class NetworkScanner {
  constructor() {
    this.deviceDiscovery = deviceDiscovery;
    this.portScanner = portScanner;
    this.networkMetrics = networkMetrics;
    this.powerManager = powerManager;
  }

  async scanNetwork(ipRange) {
    try {
      if (!ipRange) {
        ipRange = await this.deviceDiscovery.getLocalNetwork();
      }

      const [baseIp, subnet] = ipRange.split('/');
      const baseIpParts = baseIp.split('.');
      const hosts = Math.pow(2, 32 - parseInt(subnet));
      
      const devices = [];
      
      // Scan network in parallel with batching
      const batchSize = 10;
      for (let i = 0; i < hosts; i += batchSize) {
        const batch = Array.from({ length: Math.min(batchSize, hosts - i) }, (_, j) => {
          const ip = `${baseIpParts[0]}.${baseIpParts[1]}.${baseIpParts[2]}.${i + j + 1}`;
          return this.scanDevice(ip);
        });
        
        const results = await Promise.all(batch);
        devices.push(...results.filter(Boolean));
      }

      return devices;
    } catch (error) {
      console.error('Network scan error:', error);
      throw error;
    }
  }

  async scanDevice(ip) {
    const isReachable = await this.deviceDiscovery.pingHost(ip);
    if (!isReachable) return null;

    const [mac, hostname, openPorts] = await Promise.all([
      this.deviceDiscovery.getMacAddress(ip),
      this.deviceDiscovery.getHostname(ip),
      this.portScanner.scanCommonPorts(ip)
    ]);

    if (!mac && openPorts.length === 0) return null;

    return {
      ip,
      mac: mac || 'Unknown',
      name: hostname || `Device (${ip})`,
      status: 'online',
      lastSeen: new Date().toISOString(),
      openPorts
    };
  }

  async getDeviceMetrics(ip) {
    return this.networkMetrics.collectMetrics(ip);
  }

  async wakeOnLan(macAddress) {
    return this.powerManager.wakeOnLan(macAddress);
  }

  async shutdownDevice(ip, username, password) {
    return this.powerManager.shutdownDevice(ip, username, password);
  }
}

export const networkScanner = new NetworkScanner();

File: backend/python/network_utils.py
import socket
import struct
import subprocess
from typing import List, Dict
import scapy.all as scapy
import netifaces

class NetworkScanner:
    def __init__(self):
        self.interface = self._get_default_interface()

    def _get_default_interface(self) -> str:
        gateways = netifaces.gateways()
        return gateways['default'][netifaces.AF_INET][1]

    def scan_network(self, ip_range: str) -> List[Dict]:
        """
        Scan network for active devices
        Example: scanner.scan_network('192.168.1.0/24')
        """
        arp_request = scapy.ARP(pdst=ip_range)
        broadcast = scapy.Ether(dst="ff:ff:ff:ff:ff:ff")
        arp_request_broadcast = broadcast/arp_request
        answered_list = scapy.srp(arp_request_broadcast, timeout=1, verbose=False)[0]

        devices = []
        for element in answered_list:
            device = {
                'ip': element[1].psrc,
                'mac': element[1].hwsrc,
                'status': 'online',
                'lastSeen': 'now'
            }
            try:
                hostname = socket.gethostbyaddr(device['ip'])[0]
                device['name'] = hostname
            except:
                device['name'] = f"Unknown Device ({device['ip']})"
            
            devices.append(device)
        
        return devices

    def wake_on_lan(self, mac_address: str) -> bool:
        """
        Send Wake-on-LAN packet to device
        Example: scanner.wake_on_lan('00:11:22:33:44:55')
        """
        try:
            mac_bytes = bytes.fromhex(mac_address.replace(':', ''))
            magic_packet = b'\xff' * 6 + mac_bytes * 16
            
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
            sock.sendto(magic_packet, ('255.255.255.255', 9))
            return True
        except Exception as e:
            print(f"Error sending WOL packet: {e}")
            return False

    def shutdown_device(self, ip: str, username: str = None, password: str = None) -> bool:
        """
        Attempt to shutdown a device (requires proper authentication)
        Example: scanner.shutdown_device('192.168.1.100', 'username', 'password')
        """
        try:
            if sys.platform == 'win32':
                cmd = f'shutdown /s /m \\\\{ip} /t 0'
                if username and password:
                    # Use psexec or similar tool for authenticated shutdown
                    pass
            else:
                cmd = f'ssh {username}@{ip} "sudo shutdown -h now"'
            
            subprocess.run(cmd, shell=True, check=True)
            return True
        except Exception as e:
            print(f"Error shutting down device: {e}")
            return False